### ISSUE TITLE
Issue building MMEX with CURL on VS2017

### ISSUE THREAD

**Issue by grumpy235** [moneymanagerex/moneymanagerex#1813]
**Your OS:**
- [X] Windows
- [ ] Mac OSX
- [ ] Linux 
- [ ] Android -> please report bugs to moneymanagerex/android-money-manager-ex repo!

**MMEX version:**
- [ ] 1.4.x
- [ ] 1.3.x
- [ ] 1.2.x
- [ ] 1.1 or older
- [X] development version (git master branch)

CURL
Using the latest CURL source (7.61.1) and the instructions in BUILD.md I no longer get a static build.

The following commit on CURL appears to be the cause - https://github.com/curl/curl/commit/c892795ea3601a6d210a325b2ac566b1c30d3334

Changing DCURL_STATICLIB=ON to DBUILD_SHARED_LIBS=OFF on the cmake command appears to fix the issue

MMEX
Running the command 
	cmake -G "Visual Studio 15 2017 Win64" -DCMAKE_INSTALL_PREFIX:PATH=D:/Projects/mmex/build/install -DCMAKE_PREFIX_PATH=D:\Projects\curl-7.61.1 ..

gives the following error

-- Could NOT find compile cache tool (ccache or clcache)
-- Found CURL: D:/Projects/curl-7.61.1/lib/libcurl.lib (found version "7.61.1")
CMake Error at CMakeLists.txt:266 (SET_TARGET_PROPERTIES):
  SET_TARGET_PROPERTIES can not be used on an ALIAS target.
  
ccache and clcache are expected as not installed 

So I modified CMakeLists.txt as follows to confirm path taken..

```cmake
FIND_PACKAGE(CURL QUIET COMPONENTS libcurl CONFIG)
IF(CURL_FOUND)
    MESSAGE(STATUS "Found CURL: ${CURL_LIBRARIES} (found version \"${CURL_VERSION}\")")
    SET(CURL_VERSION_STRING ${CURL_VERSION})
ELSE()
    FIND_PACKAGE(CURL REQUIRED COMPONENTS libcurl MODULE)
    ADD_LIBRARY(CURL INTERFACE)
    TARGET_INCLUDE_DIRECTORIES(CURL INTERFACE ${CURL_INCLUDE_DIRS})
    TARGET_LINK_LIBRARIES(CURL INTERFACE ${CURL_LIBRARIES})
    ADD_LIBRARY(CURL::libcurl ALIAS CURL)
	MESSAGE(STATUS "Point A")  <---
ENDIF()
IF(WIN32 AND CURL_LIBRARIES MATCHES "${CMAKE_STATIC_LIBRARY_SUFFIX}$")
    MESSAGE(STATUS "Point B")  <---
    SET_TARGET_PROPERTIES(CURL::libcurl PROPERTIES INTERFACE_COMPILE_DEFINITIONS CURL_STATICLIB)
ENDIF()
```
and got...

-- Found CURL: D:/Projects/curl-7.61.1/lib/libcurl.lib (found version "7.61.1")
-- Point A
-- Point B
CMake Error at CMakeLists.txt:268 (SET_TARGET_PROPERTIES):
  SET_TARGET_PROPERTIES can not be used on an ALIAS target.
  
Pretty sure I have had this before but not sure how I got past it - any ideas as pretty sure the statement has been there since CURL support added.

**Comment by grumpy235** [comment]
Just found https://github.com/Adnn/curl/commit/633c055cd9e6afec1b3a19e3cb4f6fdcbdd24404 which looks like it is in the same area and sets similar target properties although my knowledge of how CMake works is almost zero

**Comment by slodki** [comment]
fix mentioned in #1692 IMHO...

**Comment by grumpy235** [comment]
Built CURL 7.61.1 as follows...

_cmake -G "Visual Studio 15 2017 Win64" -DBUILD_CURL_EXE=OFF -DHTTP_ONLY=ON -DENABLE_MANUAL=OFF -DBUILD_TESTING=OFF **-DBUILD_SHARED_LIBS=OFF** **-DCMAKE_USE_WINSSL=ON** -DCMAKE_INSTALL_PREFIX=D:\Projects\curl-7.61.1 ..
set "CL=/MP"
cmake --build . --target install --config Release --clean-first -- /maxcpucount /verbosity:minimal /nologo /p:PreferredToolArchitecture=x64_

Produced the following output messages..

LINK : warning LNK4044: unrecognized option '/MANIFEST:NO'; ignored [D:\Projects\curl-7.61.1\build\lib\libcurl.vcxproj]
  libcurl.vcxproj -> D:\Projects\curl-7.61.1\build\lib\Release\libcurl.lib
  Building Custom Rule D:/Projects/curl-7.61.1/CMakeLists.txt
  CMake does not need to re-run because D:/Projects/curl-7.61.1/build/CMakeFiles/generate.stamp is up-to-date.
  -- Install configuration: "Release"
  -- Installing: D:/Projects/curl-7.61.1/bin/curl-config
  -- Installing: D:/Projects/curl-7.61.1/lib/pkgconfig/libcurl.pc
  -- Installing: D:/Projects/curl-7.61.1/CMake/curl-config.cmake
  -- Installing: D:/Projects/curl-7.61.1/CMake/curl-config-version.cmake
  -- Installing: D:/Projects/curl-7.61.1/lib/libcurl.lib
  -- Installing: D:/Projects/curl-7.61.1/CMake/libcurl-target.cmake
  -- Installing: D:/Projects/curl-7.61.1/CMake/libcurl-target-release.cmake
  
So CURL build appears to be OK.

Commented out "ADD_LIBRARY(CURL::libcurl ALIAS CURL)" in MMEX CMakeLists and successfully created a project file with

_cmake -G "Visual Studio 15 2017 Win64" -DCMAKE_INSTALL_PREFIX:PATH=D:/Projects/mmex/build/install -DwxWidgets_ROOT_DIR="D:\Projects\wxWidgets" -DwxWidgets_LIB_DIR="D:\Projects\wxWidgets\lib\vc141_x64_dll" -DCMAKE_PREFIX_PATH=D:\Projects\curl-7.61.1 .._

(For some reason I needed DwxWidgets_ROOT_DIR and DwxWidgets_LIB_DIR otherwise I got a not found error out of CMake - perhaps because I had switched to a binary download and had not got my naming convention in line with CMake)

Ran mmex.sln in VS2017 and everything compiled fine but linkage failed as unable to find I guess some SSL externals

8>   Creating library D:/Projects/mmex/build/src/Release/mmex.lib and object D:/Projects/mmex/build/src/Release/mmex.exp
8>libcurl.lib(schannel.obj) : error LNK2019: unresolved external symbol __imp_CertOpenStore referenced in function schannel_connect_step1
8>libcurl.lib(schannel_verify.obj) : error LNK2001: unresolved external symbol __imp_CertOpenStore
8>libcurl.lib(schannel.obj) : error LNK2019: unresolved external symbol __imp_CertCloseStore referenced in function schannel_connect_step1
8>libcurl.lib(schannel_verify.obj) : error LNK2001: unresolved external symbol __imp_CertCloseStore
8>libcurl.lib(schannel.obj) : error LNK2019: unresolved external symbol __imp_CertFindCertificateInStore referenced in function schannel_connect_step1
8>libcurl.lib(schannel.obj) : error LNK2019: unresolved external symbol __imp_CertFreeCertificateContext referenced in function schannel_connect_step1
8>libcurl.lib(schannel_verify.obj) : error LNK2001: unresolved external symbol __imp_CertFreeCertificateContext
8>libcurl.lib(schannel.obj) : error LNK2019: unresolved external symbol __imp_CryptStringToBinaryA referenced in function schannel_connect_step1
8>libcurl.lib(schannel_verify.obj) : error LNK2019: unresolved external symbol __imp_CertAddCertificateContextToStore referenced in function add_certs_to_store
8>libcurl.lib(schannel_verify.obj) : error LNK2019: unresolved external symbol __imp_CertGetNameStringA referenced in function verify_certificate
8>libcurl.lib(schannel_verify.obj) : error LNK2019: unresolved external symbol __imp_CryptQueryObject referenced in function add_certs_to_store
8>libcurl.lib(schannel_verify.obj) : error LNK2019: unresolved external symbol __imp_CertCreateCertificateChainEngine referenced in function verify_certificate
8>libcurl.lib(schannel_verify.obj) : error LNK2019: unresolved external symbol __imp_CertFreeCertificateChainEngine referenced in function verify_certificate
8>libcurl.lib(schannel_verify.obj) : error LNK2019: unresolved external symbol __imp_CertGetCertificateChain referenced in function verify_certificate
8>libcurl.lib(schannel_verify.obj) : error LNK2019: unresolved external symbol __imp_CertFreeCertificateChain referenced in function verify_certificate
8>D:\Projects\mmex\build\install\bin\mmex.exe : fatal error LNK1120: 12 unresolved externals

The number of hits on Google for the above suggests the problem has been seen lots of times before but I am unclear on how to fix it. Some early reports suggest crypt32 missing but this has been fixed by the CURL team and is confirmed below..

CMakeCache.txt for 7.61.0 contains...
libcurl_LIB_DEPENDS:STATIC=general;winmm;general;ws2_32;general;advapi32;general;crypt32;
and so does 7.61.1

Any ideas?

**Comment by grumpy235** [comment]
After digging around on Google found a suggestion for the problem..
#pragma comment(lib, "crypt32.lib")
Added this to constants.h, re-compiled and now have an executable with CURL 7.61.1 which runs although I have not checked out CURL functionality.
I'm guessing that there may be a neater way of solving this perhaps by a compiler parameter at CMake time but my detailed knowledge of CMake and the C++ compiler is still sketchy - understand the concepts as they have not changed much in years but my implementation knowledge is poor :confused:

**Comment by slodki** [comment]
The following solution:

```cmake
# --------- curl dependency ---------
# curl version >=7.57 can have config files
FIND_PACKAGE(CURL QUIET COMPONENTS libcurl CONFIG)
IF(CURL_FOUND)
    MESSAGE(STATUS "Found CURL: ${CURL_LIBRARIES} (found version \"${CURL_VERSION}\")")
    GET_TARGET_PROPERTY(LIBCURL_TYPE CURL::libcurl TYPE)
    # CURL_STATICLIB is missing for curl versions <7.61.1
    IF(CURL_VERSION VERSION_LESS 7.61.1 AND LIBCURL_TYPE STREQUAL STATIC_LIBRARY)
        SET_TARGET_PROPERTIES(CURL::libcurl PROPERTIES
            INTERFACE_COMPILE_DEFINITIONS CURL_STATICLIB)
    ENDIF()
    SET(CURL_VERSION_STRING ${CURL_VERSION})
ELSE()
    FIND_PACKAGE(CURL REQUIRED MODULE)
    # CMake version <3.12 doesn't create CURL::libcurl target
    IF(NOT TARGET CURL::libcurl)
        ADD_LIBRARY(CURL::libcurl IMPORTED)
        SET_TARGET_PROPERTIES(CURL::libcurl PROPERTIES
            INTERFACE_INCLUDE_DIRECTORIES "${CURL_INCLUDE_DIRS}"
            IMPORTED_LINK_INTERFACE_LANGUAGES C
            IMPORTED_LOCATION "${CURL_LIBRARIES}"
            IMPORTED_CONFIGURATIONS RELEASE
            IMPORTED_LOCATION_RELEASE "${CURL_LIBRARIES}")
    ENDIF()
    IF(CURL_LIBRARIES MATCHES "${CMAKE_STATIC_LIBRARY_SUFFIX}$")
        SET_TARGET_PROPERTIES(CURL::libcurl PROPERTIES INTERFACE_COMPILE_DEFINITIONS CURL_STATICLIB)
    ENDIF()
ENDIF()
```
will work. But there are bugs in FindwxWidgets.cmake module from CMake - it prevents linking static libcurl <7.57. If you use the latest it will work.

**Comment by grumpy235** [comment]
Tried the above with latest CMake and CMakeLists now runs clean so that fixes the ALIAS issue - many thanks.

I still get the same bunch of linkage messages when running the generated mmex.sln unless I add the following to the mmex source
#pragma comment(lib, "crypt32.lib")

I guess this is a Windows unique issue as Linux will be packaged totally differently ?

**Comment by grumpy235** [comment]
Added the following to constants.h
`#ifdef _WIN32
	#pragma comment(lib, "crypt32.lib")
#endif`
Now Linker completes and I get a mmex module with CURL 7.61.1
Clicked on the "Report a Bug" link and I get to the correct webpage so I guess it is working ?

**Comment by slodki** [comment]
I you use latest curl version >=7.57 it must compile without manual fixes.
Curl installation should create *.cmake configuration files:
```
  -- Install configuration: "Release"
  -- Installing: C:/curl-7_57_0/vc141_x64_Release_dll/bin/curl-config
  -- Installing: C:/curl-7_57_0/vc141_x64_Release_dll/lib/pkgconfig/libcurl.pc
  -- Installing: C:/curl-7_57_0/vc141_x64_Release_dll/include/curl
  -- Installing: C:/curl-7_57_0/vc141_x64_Release_dll/include/curl/curl.h
  -- Installing: C:/curl-7_57_0/vc141_x64_Release_dll/include/curl/curlver.h
  -- Installing: C:/curl-7_57_0/vc141_x64_Release_dll/include/curl/easy.h
  -- Installing: C:/curl-7_57_0/vc141_x64_Release_dll/include/curl/mprintf.h
  -- Installing: C:/curl-7_57_0/vc141_x64_Release_dll/include/curl/multi.h
  -- Installing: C:/curl-7_57_0/vc141_x64_Release_dll/include/curl/stdcheaders.h
  -- Installing: C:/curl-7_57_0/vc141_x64_Release_dll/include/curl/system.h
  -- Installing: C:/curl-7_57_0/vc141_x64_Release_dll/include/curl/typecheck-gcc.h
  -- Installing: C:/curl-7_57_0/vc141_x64_Release_dll/CMake/curl-config.cmake
  -- Installing: C:/curl-7_57_0/vc141_x64_Release_dll/CMake/curl-config-version.cmake
  -- Installing: C:/curl-7_57_0/vc141_x64_Release_dll/lib/libcurl.lib
  -- Installing: C:/curl-7_57_0/vc141_x64_Release_dll/CMake/libcurl-target.cmake
  -- Installing: C:/curl-7_57_0/vc141_x64_Release_dll/CMake/libcurl-target-release.cmake
```
with all required linker parameters inside. If some linker params are missing you don't use cmake config files (not installed)?

**Comment by slodki** [comment]
Report a Bug doesn't use curl - it just start external browser with url passed as string, no http connection is created by mmex.
Curl is used for downloading currency rates, check latest mmex version, communicate with webapp...

**Comment by thinkpad4** [comment]
I'm getting a error with the latest nightly on windows 7, libcurl.dll missing.

**Comment by slodki** [comment]
Nightly is broken, fix is ready but i'm waiting for complete solution.

**Comment by grumpy235** [comment]
@slodki,
I have all the files you mention above, slightly different higher directory structure, but base folder(s) are the same. I have tried to compare my 7.61.1 folder with the 7.61.0 folder (this links cleanly) and they appear to match both in directory structure and member contents.
I suppose it is possble that Curl 7.61.1 has changed something in the WINSSL support?

Unfortunately I do not use currency or the webapp but "check for updates" returns "You already have the latest version" so does this mean Curl is working?

**Comment by slodki** [comment]
There should be something like this:
```cmake
# Create imported target CURL::libcurl
add_library(CURL::libcurl STATIC IMPORTED)

set_target_properties(CURL::libcurl PROPERTIES
  INTERFACE_COMPILE_DEFINITIONS "CURL_STATICLIB"
  INTERFACE_INCLUDE_DIRECTORIES "${_IMPORT_PREFIX}/include"
  INTERFACE_LINK_LIBRARIES "winmm;ws2_32;advapi32;crypt32"
)
```
in libcurl-target.cmake. As you can see LINK_LIBRARIES contains all required libraries for this compilation. This file is used for MMEX build. If you must manually add this parameters you must not use this file. You can check this by changing MMEX CMakeLists to:
```cmake
# --------- curl dependency ---------
# curl version >=7.57 can have config files
FIND_PACKAGE(CURL QUIET COMPONENTS libcurl CONFIG)
IF(CURL_FOUND)
    MESSAGE(STATUS "Found CURL: ${CURL_LIBRARIES} (found version \"${CURL_VERSION}\")")
    GET_TARGET_PROPERTY(LIBCURL_TYPE CURL::libcurl TYPE)
    # CURL_STATICLIB is missing for curl versions <7.61.1
    IF(CURL_VERSION VERSION_LESS 7.61.1 AND LIBCURL_TYPE STREQUAL STATIC_LIBRARY)
        SET_TARGET_PROPERTIES(CURL::libcurl PROPERTIES
            INTERFACE_COMPILE_DEFINITIONS CURL_STATICLIB)
    ENDIF()
    SET(CURL_VERSION_STRING ${CURL_VERSION})
ELSE()
    MESSAGE(FATAL_ERROR "libcurl .cmake config file not found!")
ENDIF()
```
You will be warned if libcurl configuration cannot be imported

**Comment by grumpy235** [comment]
Thanks for the suggestions.
I can see ibcurl-target.cmake in my D:\Projects\curl-7.61.1\CMake and it contains exactly what you quote above

Made your suggested change to CMakeLists for mmex and got the following...

CMake Error at CMakeLists.txt:267 (MESSAGE):
  libcurl .cmake config file not found!  **<-----**
  
Changed to FIND_PACKAGE(CURL **REQUIRED** COMPONENTS libcurl CONFIG)

and now get..

CMake Error at C:/Program Files/CMake/share/cmake-3.12/Modules/FindPackageHandleStandardArgs.cmake:137 (message):
  Could NOT find OpenSSL, try to set the path to OpenSSL root folder in the
  system variable OPENSSL_ROOT_DIR (missing: OPENSSL_CRYPTO_LIBRARY
  OPENSSL_INCLUDE_DIR)
Call Stack (most recent call first):
  C:/Program Files/CMake/share/cmake-3.12/Modules/FindPackageHandleStandardArgs.cmake:378 (_FPHSA_FAILURE_MESSAGE)
  C:/Program Files/CMake/share/cmake-3.12/Modules/FindOpenSSL.cmake:412 (find_package_handle_standard_args)
  C:/Program Files/CMake/share/cmake-3.12/Modules/CMakeFindDependencyMacro.cmake:48 (find_package)
  D:/Projects/curl-7.61.1/CMake/curl-config.cmake:14 (find_dependency)
  CMakeLists.txt:256 (FIND_PACKAGE)
  
So goes looking for OpenSSL, my command used to build was
cmake -G "Visual Studio 15 2017 Win64" -DBUILD_CURL_EXE=OFF -DHTTP_ONLY=ON -DENABLE_MANUAL=OFF -DBUILD_TESTING=OFF -DBUILD_SHARED_LIBS=OFF **-DCMAKE_USE_WINSSL=ON** -DCMAKE_INSTALL_PREFIX=D:\Projects\curl-7.61.1 ..

Extract from D:\Projects\curl-7.61.1\build\CMakeCache.txt....

//Use OpenSSL code. Experimental
CMAKE_USE_OPENSSL:BOOL=OFF

//enable Windows native SSL/TLS
CMAKE_USE_WINSSL:BOOL=ON

and

//Dependencies for the target
libcurl_LIB_DEPENDS:STATIC=general;winmm;general;ws2_32;general;advapi32;general;crypt32;

Looks like my modifications to the original Curl Cmake command are not correct or problem with delivered Curl Cmake code ?

**Comment by grumpy235** [comment]
Definitely something wrong with my Curl build. 
Found the following.

**D:\Projects\curl-7.61.0\build\lib\libcurl.sln** contains a Link section for each of the configs containing all the required Windows dlls including crypt32

**D:\Projects\curl-7.61.1\build\lib\libcurl.sln** no Link sections at all.

Which I guess ties in with why I get the OPEN_SSL reference above?

**Comment by slodki** [comment]
If you got *libcurl .cmake config file not found!* message it means you must verify `-DCMAKE_PREFIX_PATH=c:\libcurl` parameter while configuring MMEX. It's responsible to include libcurl configuration into MMEX buil environment.

Please remember - binary distributions of curl don't include *.cmake config for external projects and can be configured with different options - it's required to read their documentation to prepare correct linking. We use our own compilation of libcurl in MMEX to simplify integration.

**Comment by grumpy235** [comment]
I understand. 
Curl 7.61.0 works links just fine with latest Mmex source so pretty sure the fault lies either with a change made to Curl at 7.61.1 level or I have the build command wrong. I'll have a play tomorrow.
Just for interest do your Nightly builds of Mmex for Windows work with Curl 7.61.1

**Comment by slodki** [comment]
https://gitlab.kitware.com/cmake/cmake/issues/18378

**Comment by grumpy235** [comment]
Just got back to this and looks like it is a problem with the CURL CMake build (they do warn about it being poorly maintained!)
Building 7.61.0 with the instructions in mmex BUILD.md then first few lines of D:\Projects\curl-7.61.0\Make\curl-config.cmake are

****
get_filename_component(_DIR "${CMAKE_CURRENT_LIST_FILE}" PATH)

if(NOT CURL_FIND_COMPONENTS)
    set(CURL_FIND_COMPONENTS curl libcurl)
    if(CURL_FIND_REQUIRED)
        set(CURL_FIND_REQUIRED_curl TRUE)
        set(CURL_FIND_REQUIRED_libcurl TRUE)
    endif()
endif()

set(_curl_missing_components)
****

Building 7.61.1 with the following command..

cmake -G "Visual Studio 15 2017 Win64" -DBUILD_CURL_EXE=OFF -DHTTP_ONLY=ON -DENABLE_MANUAL=OFF -DBUILD_TESTING=OFF -DBUILD_SHARED_LIBS=OFF -DCMAKE_USE_OPENSSL=OFF -DCMAKE_USE_WINSSL=ON -DCMAKE_INSTALL_PREFIX=D:\Projects\curl-7.61.1 ..

The first few lines of curl-config.cmake are..

****
get_filename_component(_DIR "${CMAKE_CURRENT_LIST_FILE}" PATH)

if(NOT CURL_FIND_COMPONENTS)
  set(CURL_FIND_COMPONENTS curl libcurl)
  if(CURL_FIND_REQUIRED)
    set(CURL_FIND_REQUIRED_curl TRUE)
    set(CURL_FIND_REQUIRED_libcurl TRUE)
  endif()
endif()

**include(CMakeFindDependencyMacro)
if(CURL_FIND_REQUIRED_libcurl)
    find_dependency(OpenSSL "")
endif()**

set(_curl_missing_components)
****

So for some reason, it thinks OpenSSL is required perhaps caused by the BUILD_SHARED_LIBS commit [link](https://github.com/curl/curl/commit/c892795ea3601a6d210a325b2ac566b1c30d3334#diff-af3b638bc2a3e6c650974192a53c7291) and these few extra lines appear to cause all the pain.
Commenting out the extra lines and mmex builds just fine with your 'Curl Dependency code' from above

Spent some time trying to find how the OpenSSL requirement gets generated and my lack of CMake knowledge is proving a problem

**Comment by slodki** [comment]
Sorry, but tested with 7.61.1 and 7.57, both static and dynamic libcurl builds on Windows using clean appveyor environment and all is configured ok and MMEX is build without errors with following code in CMakeFiles.txt:
```cmake
# --------- curl dependency ---------
# curl version >=7.57 can have config files
FIND_PACKAGE(CURL QUIET COMPONENTS libcurl CONFIG)
IF(CURL_FOUND)
    MESSAGE(STATUS "Found CURL: ${CURL_LIBRARIES} (found version \"${CURL_VERSION}\")")
    GET_TARGET_PROPERTY(LIBCURL_TYPE CURL::libcurl TYPE)
    # CURL_STATICLIB is missing for curl versions <7.61.1
    IF(CURL_VERSION VERSION_LESS 7.61.1 AND LIBCURL_TYPE STREQUAL STATIC_LIBRARY)
        SET_TARGET_PROPERTIES(CURL::libcurl PROPERTIES
            INTERFACE_COMPILE_DEFINITIONS CURL_STATICLIB)
    ENDIF()
    SET(CURL_VERSION_STRING ${CURL_VERSION})
ELSE()
    MESSAGE(FATAL_ERROR "libcurl .cmake config file not found!")
ENDIF()
```

The problem is with libcurl <7.57 or compiled using other tools like autotools etc. where *.cmake config files are missing. This problem is described in CMake issue linked above.

**Comment by grumpy235** [comment]
Thanks for the update.
Maybe I'll play with Appveyor might give me a hint as to what is wrong with my Curl CMake command. Was going to run Curl CMake with --trace and --debug (now I have read the manual!) but I guess the output will be vast and may not get me much further forward.

Just a thought does your Windows/Curl 7.61.1 curl-config.cmake contain find_dependency(OpenSSL "") and if so do you have OpenSSL installed on your Windows ?

Guess this issue can be closed as looks like problem is unique to me, at least it has got my little grey cells ticking over though!

**Comment by grumpy235** [comment]
Sorry can't seem to let this drop.
Think I have found the curl [commit](https://github.com/curl/curl/commit/7867aaa9a01decf93711428462335be8cef70212) that is causing my issue.
This simply drops the following code

include(CMakeFindDependencyMacro)
if(CURL_FIND_REQUIRED_libcurl)
    find_dependency(OpenSSL "@OPENSSL_VERSION_MAJOR@")
endif()

into CMake/curl-config.cmake. This is fine on any system which contains an OpenSSL package, almost certainly the vast majority but will not work for a vanilla Windows system.
My Curl CMakeCache contains

//Use OpenSSL code. Experimental
CMAKE_USE_OPENSSL:BOOL=OFF

So bracketing the above code with if(CMAKE_USE_OPENSSL)...endif() fixes my problem

**Comment by slodki** [comment]
see curl/curl#3001 or curl/curl#2849

**Comment by grumpy235** [comment]
Many thanks was beginning to think I was imagining my problem  :smile:

**Comment by grumpy235** [comment]
Just tried the Curl build with your mod to curl-config.cmake.in which gets copied to curl-config.cmake at the end of CMakeLists with @ONLY option.
I end up with 
      if( AND CURL_FIND_REQUIRED_libcurl)
and this throws an error in the Mmex build as of course it is invalid syntax.

**Comment by slodki** [comment]
Fixed in cURL 7.62

**Comment by grumpy235** [comment]
@slodki 
I noticed the Appveyor build failed with CURL 7.62 giving exactly the same error as I had when attempting to build on my PC

[00:08:55] CMake Warning at CMakeLists.txt:256 (FIND_PACKAGE):
[00:08:55]   Found package configuration file:
[00:08:55] 
[00:08:55]     C:/curl-7_62_0/vc141_x64_Release_lib/lib/cmake/CURL/CURLConfig.cmake
[00:08:55] 
[00:08:55]   but it set CURL_FOUND to FALSE so package "CURL" is considered to be NOT
[00:08:55]   FOUND.

It would appear that CURL commit 6932849 "CMake: Improve config installation" is probably the culprit.
After this change then ..\curl-7.62.0\lib\cmake\CURL\CURLConfig.cmake contains

macro(check_required_components _NAME)
  foreach(comp ${${_NAME}_FIND_COMPONENTS})
    if(NOT ${_NAME}_${comp}_FOUND)
      if(${_NAME}_FIND_REQUIRED_${comp})
        set(${_NAME}_FOUND FALSE)
      endif()
    endif()
  endforeach()
endmacro()

Which compares with the following from curl-config.cmake.in for 7.61.1

foreach(_comp ${CURL_FIND_COMPONENTS})
  if(EXISTS "${_DIR}/${_comp}-target.cmake")
    include("${_DIR}/${_comp}-target.cmake")
    set(CURL_${_comp}_FOUND TRUE)
  else()
    set(CURL_${_comp}_FOUND FALSE)
    if(CURL_FIND_REQUIRED_${_comp})
      set(CURL_FOUND FALSE)
      list(APPEND _curl_missing_components ${_comp})
    endif()
  endif()
endforeach()

I got round this by copying the 7.61.1 Cmake code into the 7.62 Cmake macro as I could not get my head round what needed changing in the Curl Cmake code to make it work without the patch.

**Comment by slodki** [comment]
fixed in #1875
